lib/flox-env.nix

```diff
@@ -3,62 +3,91 @@
-# User API
-pkgs': toml: pins: let
-  tie = {
-    pkgs = pkgs';
-    mach = floxpkgs.inputs.mach-nix.lib.${pkgs.system};
-    vscodeLib = lib.vscode;
-  };
-  data = {
-    func = floxEnv;
-    attrs = if builtins.isAttrs toml then toml else builtins.fromTOML (builtins.readFile toml);
-  };
-  pkgs = tie.pkgs;
+}: let
+  floxpkgs = self;
+in
+  # User API
+  pkgs': toml: pins: let
+    tie = {
+      pkgs = pkgs';
+      mach = floxpkgs.inputs.mach-nix.lib.${pkgs.system};
+      vscodeLib = lib.vscode;
+    };
+    data = {
+      func = floxEnv;
+      attrs =
+        if builtins.isAttrs toml
+        then toml
+        else builtins.fromTOML (builtins.readFile toml);
+    };
+    pkgs = tie.pkgs;
     floxEnv = {programs, ...}: let
-    python = let
-      mach = import (floxpkgs.inputs.mach-nix + "/default.nix") {
-        inherit pkgs;
-        dataOutdated = false;
-        pypiData = floxpkgs.inputs.mach-nix.inputs.pypi-deps-db;
-      };
-    in
-      mach.mkPython (programs.python
-        // {
-          ignoreDataOutdated = true;
-        });
-    paths = let
-      handler = {
-        python = python;
-        vscode =
-          floxpkgs.lib.vscode.configuredVscode
-          pkgs
-          programs.vscode
-          pins.vscode-extensions;
+      python = config: let
+        mach = import (floxpkgs.inputs.mach-nix + "/default.nix") {
+          inherit pkgs;
+          dataOutdated = false;
+          pypiData = floxpkgs.inputs.mach-nix.inputs.pypi-deps-db;
+        };
+      in
+        mach.mkPython (config
+          // {
+            ignoreDataOutdated = true;
+          });
+      paths = let
+        handler = {
+          python = python;
+          vscode = config:
+            floxpkgs.lib.vscode.configuredVscode
+            pkgs
+            config
+            pins.vscode-extensions;
 
-        # insert excpetions here
-        __functor = self: key: attr:
-          self.${key}
-          or (
-            if attr ? version
-            then "${key}@${attr.version}"
-            else pkgs.${key}
-          );
-      };
+          catalog = programs: let
+            defaultVersion = with builtins;
+              path: let
+                # TODO shouldn't include recurseForDerivations
+                versionsWithUnderScores = attrNames (lib.getAttrFromPath
+                  path
+                  nixpkgs-flox.catalog.${pkgs.system});
+                versions = map (version: replaceStrings ["_"] ["."] version) versionsWithUnderScores;
+                sortedVersions = sort (v1: v2: (compareVersions v1 v2) == 1) versions;
+              in
+                replaceStrings ["."] ["_"] (head sortedVersions);
+            ensureStabilityAndVersion = with builtins;
+              path:
+                if length path == 2
+                then path ++ [(defaultVersion path)]
+                else if length path == 1
+                then ensureStabilityAndVersion (path ++ ["stable"])
+                else path;
```

Comment:
Could drop this chunk if we had passthu setup as described in https://github.com/flox/update-catalog/issues/4